Material Date Picker redesign

darrenaustin opened the pull request:
## Description

The design for the Material Date Picker has been updated:

https://material.io/components/pickers/#mobile-pickers

This PR updates Flutter's implementation to match the new design.

![DatePickerM2](https://user-images.githubusercontent.com/19588/75818324-8432ad00-5d4d-11ea-9633-d322b1d34b5b.gif)

In addition to just updating the visuals and controls, a new text input mode was added to the dialog as per the design.

The API changes are documented at: https://flutter.dev/go/material-date-picker-redesign

Note that this does not include support for the Date Range picker as that will be coming in a future PR.

Also, credit to @rami-a who implemented a similar internal design that I based this on. However, any problems were introduced by me, so don't blame him :).

## Related Issues

* Fixes #50133

## Tests

The tests in `date_picker_test.dart` have been updated to test the new UI implementation. Tests for the new text input mode were also added to cover that.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

darrenaustin committed:
Date Picker redesign

Updated the showDatePicker() implementation to use the latest Material design UI.

rami-a commented on packages/flutter/test/material/date_picker_test.dart:
```diff
@@ -150,672 +70,850 @@ void _tests() {
-      await tester.tap(find.text('17')); // Odd, doesn't work.
-      await tester.tap(find.text('OK'));
-      expect(await date, equals(DateTime(2017, DateTime.january, 10)));
+    testWidgets('Can select initial date picker mode', (WidgetTester tester) async {
```
This test appears to be a duplicate of an earlier one in here

rami-a reviewed (commented):
Looks mostly good, comments are pretty minor but noticed a couple of small bugs when testing it out.

- Header overflows in landscape mode with text input and keyboard up
- Landscape mode should probably cap the width (especially in text input mode), otherwise it looks very wide.
- Rotating the device does not maintain the current month the user is viewing, it goes back to the initial month.

darrenaustin committed:
Added a key for the CalendarDatePicker so that the dialog will maintain state between orientation changes.

darrenaustin committed:
Found a better way to reset the screen test size between tests.

darrenaustin committed:
Doc and other small updates from code review.

darrenaustin committed:
Fixed the size of the dialog so it doesn't layout weird on large displays.

darrenaustin committed:
Fixed the landscape input dialog to handle a small size with the keyboard up.

darrenaustin commented on packages/flutter/lib/src/material/pickers/date_utils.dart:
```diff
@@ -0,0 +1,117 @@
+import '../material_localizations.dart';
+
+/// Returns a [DateTime] with just the date of the original, but no time set.
+DateTime dateOnly(DateTime date) {
```
I thought about that, but I am bit ambivalent about extension methods. They do make things easier to read, but it is not obvious that they aren't part of the main object's API, which can lead to confusion. Given that these are all internal anyway, I just left them as util functions.

darrenaustin commented on packages/flutter/lib/src/material/pickers/input_date_picker.dart:
```diff
@@ -0,0 +1,279 @@
+            SizedBox(height: orientation == Orientation.portrait ? 16 : 26),
+            TextFormField(
+              decoration: InputDecoration(
+                border: const UnderlineInputBorder(),
```
My hope is that we can expose these kinds of things in a DatePickerTheme pass.

darrenaustin commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,433 @@
+        return null;
+      }),
+      insetPadding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 24.0),
+      shape: const RoundedRectangleBorder(
```
Yup. My thought was to update the DialogTheme to match the rest of the current theme patterns, in which case we could update the defaults and then we wouldn't need to hard code those here (or just have these be fall backs when the DialogTheme didn't override them).

darrenaustin commented:
@rami-a, thanks for the review. I think I have either addressed all your feedback. There are some issues, like customizability which we will deal with in a DatePickerTheme pass in another PR.

> * Header overflows in landscape mode with text input and keyboard up
> * Landscape mode should probably cap the width (especially in text input mode), otherwise it looks very wide.
> * Rotating the device does not maintain the current month the user is viewing, it goes back to the initial month.

These should all be fixed with my latest commits. Please let me know if you see anything else.

gspencergoog commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -434,6 +475,23 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  ///
+  /// This applies the leap year logic introduced by the Gregorian reforms of
+  /// 1582. It will not give valid results for dates prior to that time.
+  int _getDaysInMonth(int year, int month) {
```
It's too bad this isn't already part of DateTime.  Maybe propose a PR for the Dart team?

darrenaustin replied:
Yeah, I will put that on my list and follow up later. There are other things like the DateTimeRange and other stuff in the date_util.dart that might be good candidates as well.

gspencergoog commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -467,6 +525,21 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
   String formatYear(DateTime date) => date.year.toString();
 
+  @override
+  String formatCompactDate(DateTime date) {
```
So what happens in other locales?  Is this just a fallback, or are the other locales just not implemented yet?

darrenaustin replied:
This is just a fallback for the default case.

Other locales are handled by the _MaterialLocalizationsDelegate:

https://github.com/flutter/flutter/pull/50546/files/4e3dc0117f2531952d92b702b2e09db2aa9f82c6#diff-945add11ecf717a6dad5cb70168cd894R625

Which will use the `intl.DateFormat.yMd` for the given locale. This is how the other formats that were already there were implemented.

gspencergoog commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+      ),
+    );
+
+    final Widget dialog = OrientationBuilder(
```
Why not just return the `OrientationBuilder` here and not bother creating `dialog` at all?

darrenaustin replied:
At one point I think I was wrapping the dialog with something else that got removed. I will simplify this. Thx.

gspencergoog commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+class _YearPickerState extends State<_YearPicker> {
+  ScrollController scrollController;
+
+  // The approximate number of years necessary to fill the available space.
```
Can you explain this a little more?  Why is it approximate, not exact, and what available space is it filling?

darrenaustin replied:
Not sure :). This was code from @rami-a that I didn't look that closely at. I'll spend some time with it and see if I can get better docs here and your next comment below.

gspencergoog commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+  void initState() {
+    super.initState();
+
+    // Set the scroll position to approximately center the initial year.
```
Again, why approximate, and when does the approximation get resolved? Is there some kind of step that narrows down the value to the exact value?

gspencergoog commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+/// calendar date picker initially appear in the [DatePickerMode.year] or
+/// [DatePickerMode.day] mode. It defaults to [DatePickerMode.day], and
+/// must be non-null.
+///
```
Extra line can be removed.

gspencergoog commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+const Size _inputLandscapeDialogSize = Size(496, 160.0);
+const Duration _dialogSizeAnimationDuration = Duration(milliseconds: 200);
+
+/// Shows a dialog containing a Material Design date picker.
```
It would be nice to add an interactive sample here sometime (not in this PR).

darrenaustin replied:
Yup. I will do this in another pass after the date range picker gets in as well.

gspencergoog commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+        case DatePickerEntryMode.calendar:
+          _entryMode = DatePickerEntryMode.input;
+          break;
+
```
nit: don't need the extra line here, IMO.

HansMuller commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+/// week. It provides controls to change the year and month that the grid is showing.
+///
+/// The calendar picker widget is rarely used directly. Instead, consider using
+/// [showDatePicker], which will create a dialog that uses this as well as provides
```
which creates a Dialog that contains a [CalendarDatePicker] and provides an optional compact view where the user can enter a date as a line of text.

HansMuller commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+        _currentDisplayedMonthDate = DateTime(date.year, date.month);
+        _previousMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, -1);
+        _nextMonthDate = utils.addMonthsToMonthDate(_currentDisplayedMonthDate, 1);
+        _pageController?.dispose();
```
This seems a bit risky. Presumably there's no way for the now-disposed page controller to be accessed, even though it's part of the current widget tree.  Similarly, I assume that there's no way for _setCurrentMonth() to be called more than once in between builds - so we will not unnecessarily dispose/create page controllers?

darrenaustin replied:
Indeed. I just refactored the page controller out of the CalendarDatePicker and into the _MonthPicker so this will no longer be an issue.

HansMuller commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+      value: widget.mode == DatePickerMode.year ? 1 : 0,
+      duration: const Duration(milliseconds: 200),
+      vsync: this,
+    )..addListener(() => setState(() {
```
Typically this is done by wrapping the `build()` widget with an AnimatedBuilder whose animation is _controller. In this case it looks like only the Transform widget needs to be wrapped with an AnimatedBuilder .. in fact maybe a RotationTransition would be better than an AnimatedBuilder

HansMuller commented on packages/flutter/lib/src/material/pickers/calendar_date_picker.dart:
```diff
@@ -0,0 +1,1050 @@
+                          child: Text(
+                            widget.title,
+                            overflow: TextOverflow.ellipsis,
+                            style: textTheme.subtitle2?.apply(
```
Do we really need to use 'apply' here? It's semantics are more complicated than copyWith

HansMuller commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+/// The returned [Future] resolves to the date selected by the user when the
+/// user confirms the dialog. If the user cancels the dialog, null is returned.
+///
+/// The [initialDate] is the initially shown and selected date in the picker.
```
This sentence is a little hard to parse.

Maybe something like: When the date picker is first displayed, [initialDate] is selected.

HansMuller commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+///   * [cancelText], label on the cancel button.
+///   * [confirmText], label on the ok button.
+///   * [errorFormatText], message used when the input text isn't in a proper date format.
+///   * [errorInvalidText], message used when the input date isn't a selectable date.
```
input date => input text

HansMuller commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+/// calendar date picker initially appear in the [DatePickerMode.year] or
+/// [DatePickerMode.day] mode. It defaults to [DatePickerMode.day], and
+/// must be non-null.
+///
```
There's no doc for the builder parameter.

darrenaustin replied:
See line 81 above.

HansMuller commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+    final MaterialLocalizations localizations = MaterialLocalizations.of(context);
+    final Orientation orientation = MediaQuery.of(context).orientation;
+    final TextTheme textTheme = theme.textTheme;
+    final double textScaleFactor = math.min(MediaQuery.of(context).textScaleFactor, 1.3);
```
This probably merits a brief explanation

HansMuller commented on packages/flutter/lib/src/material/pickers/date_picker_dialog.dart:
```diff
@@ -0,0 +1,480 @@
+      ? colorScheme.onPrimary
+      : colorScheme.onSurface;
+    final TextStyle dateStyle = orientation == Orientation.landscape
+      ? textTheme.headline5?.apply(color: dateColor)
```
Same comment as elsewhere about apply instead of copyWith

HansMuller reviewed (commented):
Mostly small stuff and wordsmithing.

darrenaustin commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -221,6 +221,29 @@ abstract class MaterialLocalizations {
+  /// Examples:
+  ///
+  /// - US English: Feb 21, 2019
+  /// - Russian: 21 февр. 2019 г.
```
No, but I used the existing date formatters the the intl package to generate them, so I assumed it would be correct 😄. That said, I will ask around and see if anyone can review these.

yjbanov replied:
I'm a native Russian speaker and I approve this format :) (caveat: I'm not an expect on proper date formatting in Russia though)

yjbanov replied:
Maybe @mraleph can double-check.

mraleph replied:
Looks fine to me, though a tad bizarre. I thought abbreviation for February uses 3 letters rather than 4, but searching around reveals that people use 4... Well, I learned something today... Maybe I was asleep on that lesson in school :)

darrenaustin commented on packages/flutter/lib/src/material/material_localizations.dart:
```diff
@@ -434,6 +475,23 @@ class DefaultMaterialLocalizations implements MaterialLocalizations {
+  /// Returns the number of days in a month, according to the proleptic
+  /// Gregorian calendar.
+  ///
+  /// This applies the leap year logic introduced by the Gregorian reforms of
```
Yeah, this was from the original date_picker code. As Greg mentioned above, it might be good to add this to Dart's date package.

darrenaustin commented on packages/flutter/lib/src/material/pickers/input_date_picker.dart:
```diff
@@ -0,0 +1,280 @@
+
+  final WhitelistingTextInputFormatter _filterFormatter =
+    // Only allow digits and separators (slash, dot, comma, hyphen, space).
+    WhitelistingTextInputFormatter(RegExp(r'[\d\/\.,-\s]+'));
```
There is already one above:

```dart
                // TODO(darrenaustin): localize date separator '/'
                _DateTextInputFormatter('/'),
```

These separators in the regex are any separators that the keyboard might be supplying that we always want to just map to the locale's separator. Probably not necessary, but I was worried about a date keyboard that comes up might not have locale's separator on it. The user would still be able to use these other separator keys and they would map to the right one.

darrenaustin commented on packages/flutter/lib/src/material/pickers/date_utils.dart:
```diff
@@ -0,0 +1,117 @@
+/// year.
+bool isSameDay(DateTime dateA, DateTime dateB) {
+  return
+    dateA.year == dateB.year &&
```
Yeah, the `DateTime` constructor handles any wrapping needed for the day and month fields.

darrenaustin committed:
Review comment updates
- Minor documentation fixes
- Switching to copyWith instead of apply for text styles
- Optimization in getDaysInMonth

darrenaustin committed:
More review updates:
- Simplified the dialog build method.
- Replaced the Transform widget with a RotationTransition widget for the year/day mode drop down arrow animation.

darrenaustin committed:
Refactored the page controller out of the CalendarDatePicker and into the _MonthPicker to cleanup the controller handling.

darrenaustin committed:
Removed some extraneous end-of-line spaces.

darrenaustin committed:
Removed some extraneous end-of-line spaces.

darrenaustin committed:
Adding missing dispose of the page controller in _MonthPicker

darrenaustin commented:
> Mostly small stuff and wordsmithing.

Thanks for the review @HansMuller. Can you take another look with my latest commits?

HansMuller reviewed (approved):
LGTM

Hixie commented:
Do we know if this will address https://github.com/flutter/flutter/issues/49809 or https://github.com/flutter/flutter/issues/49810?

darrenaustin commented:
> Do we know if this will address #49809 or #49810?

Unfortunately, this PR will not address those. They are still on my list. After I land the date range picker I will update them to address #49809. The Time Picker is currently being update by @rami-a and we can look at #49810 after he is finished with that.

AndroidNils commented:
Thanks for this great PR, @darrenaustin! For our web app we are on master channel. Yesterday we pushed a new web app update and today we received the first user feedbacks that they love the new date picker 😍 
 
![image](https://user-images.githubusercontent.com/24459435/77646464-e5245e00-6f64-11ea-96a6-c89e9cd5c358.png)

https://web.sharezone.net

ilechcod commented:
Pls is the new material date picker available for use on flutter now? Do I need to update my current flutter version?

Hixie commented on packages/flutter/lib/src/material/pickers/input_date_picker.dart:
```diff
@@ -0,0 +1,281 @@
+      text: filteredValue.text.replaceAll(RegExp(r'[\D]'), separator),
+    );
+  }
+}
```
@darrenaustin I'm about to check in something that replaces WhitelistingTextInputFormatter and BlacklistingTextInputFormatter with a single FilteringTextInputFormatter that handles replacement even for the case of allow-list filtering, which you might be able to use here.

Hixie commented on packages/flutter/lib/src/material/pickers/date_picker_header.dart:
```diff
@@ -0,0 +1,193 @@
+// NOTE: This is an internal implementation file. Even though there are public
+// classes and functions defined here, they are only meant to be used by the
+// date picker implementation and are not exported as part of the Material library.
+// See pickers.dart for exactly what is considered part of the public API.
```
As a general rule, we should just export everything and support everything as a public API. The reality is that everyone can see everything we do, so they'll just import the files directly if they want to, so we have to consider everything public anyway. I think we should just make everything public rather than start having semi-private APIs like this.

Hixie commented on packages/flutter/lib/src/material/pickers/date_picker_header.dart:
```diff
@@ -0,0 +1,193 @@
+import '../text_theme.dart';
+import '../theme.dart';
+
+// NOTE: This is an internal implementation file. Even though there are public
```
trivial nit: "NOTE" is against our style guide. :-)
